### PR TITLE
[Estuary] Refactor mediaflags and metadata variables

### PR DESCRIPTION
**Pull request by jjd-uk**
## Description
Refector media flags to be more flexible by removing the use of PNG image files, and to instead use a single standard image flag overlaid with a string provided by a set of skin variables for each type of media flag.

## Motivation and context
Currently mediaflags use image files to convey textual information. Thus everytime a new flag is needed it requires a new PNG image file to be created. This has created inconsistancies as few people seem to be aware that the flags date from the orignal pre-master merger of Estuary which used the Lato font, this has meant flags have been added with a mix of fonts/styles.

We already have the concept of using a empty rectangle image PNG (https://github.com/xbmc/xbmc/blob/master/addons/skin.estuary/media/flags/flag.png) for the duration flag, see https://github.com/xbmc/xbmc/blob/master/addons/skin.estuary/xml/Includes.xml#L384-L398

So this is based on a slighlty modified version of this method to provide flags for all metadata.

**IMPROVEMENTS**

So the improvements this PR brings are:

**1.** All flags use the skin font.

**2**. Keep the concept of formatting the string returned, for example instead of eac3 then a more user friendly Dolby D+ is returned, this is done by using skin Variables rather than PNG image files.

**3.** The use of skin variables allow the labels to be easily changed so:

- It's a simple xml file edit to the variable if a user wants change what is display, for example EAC3 displays **Dolby D+** but a user may prefer **DDP**

- When a new codec is added, then if the ListItem, e.g. ListItem.AudioCodec, is filled by the core core adding the new codec then the skin will automatically show the codec as the variables will fallback to the unformatted ListItem e.g. $INFO[ListItem.AudioCodec]. 

- Simplifies the addition of new formatted flag, for example when FFMPEG 7 is merged it should be easy add formated flags for Atmos and DTS:X.

**4.** When Video Versions was introduced the Choose and Manage dialogs used variables to provide metadata info to help distinguish versions. So now a common set of variables are used for both mediaflags and this versions/extras metadata.

**CHANGES**

I've taken the opportunity with this to update the labels in some cases, for example:

For DTS-HD MA the image flag contained DTS-MA, so this has been changed to a label of **DTSHD-MA**.
For Dolby TrueHD the image flag contained DOLBY-HD so this has been changed to a label of **TrueHD**.
For EAC3  the image flag contained DOLBY so this has been changed to a label of **Dolby D+**.

## How has this been tested?
Been tested for the past month on my Windows test/code PC using a variety of SD, HD, UHD, HDR videos and a variety music files.

## Screenshots (if appropriate):
The look of the media flags should be almost identical as before.

image

image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: addons/skin.estuary/xml/Variables.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Variables.xml:812]: Couldn't this be simplified?
```
	<variable name="VideoListAspectVar">
		<value condition="String.IsEqual(ListItem.VideoAspect,1.33)">4:3</value>
		<value condition="String.IsEqual(ListItem.VideoAspect,1.78)">16:9</value>
		<value>$INFO[ListItem.VideoAspect,,:1]</value>
	</variable>
```

Seems a little inconsistent to add  :1 for all besides these 2 but I assume it's just due to what people expect to see regarding 4 by 3 and 16 by 9

Same with below:

```
	<variable name="VideoPlayerAspectVar">
		<value condition="String.IsEqual(VideoPlayer.VideoAspect,1.33)">4:3</value>
		<value condition="String.IsEqual(VideoPlayer.VideoAspect,1.78)">16:9</value>
		<value>$INFO[ListItem.VideoAspect,,:1]</value>
	</variable>
```

File: addons/skin.estuary/xml/Variables.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Variables.xml:827]: VideoPlayer.VideoAspect

File: addons/skin.estuary/xml/Includes.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Includes.xml:438]: VideoPlayer.IsStereoscopic

File: addons/skin.estuary/xml/Includes.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Includes.xml:391]: Is $PARAM[infolabel_prefix] still needed? I don't see it referenced anywhere.

**Comment by MikeSiLVO** [comment]
Not sure if better or worse but what about something like the below?

```xml
	<include name="MediaFlags">
		<control type="grouplist">
			<orientation>horizontal</orientation>
			<right>20</right>
			<bottom>0</bottom>
			<height>60</height>
			<align>right</align>
			<itemgap>10</itemgap>
			<width>1900</width>
			<usecontrolcoords>true</usecontrolcoords>
			<include content="MediaFlagsLibrary" condition="!Window.IsActive(fullscreenvideo)">
				<param name="visible">![Container.Content(albums) | Container.Content(songs)]</param>
			</include>
			<include content="MediaFlagsVideoPlayer" condition="Window.IsActive(fullscreenvideo)">
				<param name="visible">[Player.ShowInfo | Window.IsActive(fullscreeninfo)] + !Player.ChannelPreviewActive + Window.IsActive(fullscreenvideo)</param>
			</include>
		</control>
	</include>
	<include name="MediaFlagsLibrary">
		<visible>$PARAM[visible]</visible>
		<control type="group">
			<width>150</width>
			<visible>System.AddonIsEnabled(resource.images.studios.white) + !String.IsEmpty($PARAM[infolabel_prefix]ListItem.Studio)</visible>
			<include content="MediaFlag">
				<param name="texture" value="$INFO[$PARAM[infolabel_prefix]ListItem.Studio,resource://resource.images.studios.white/,.png]" />
			</include>
		</control>
		<control type="group">
			<top>10</top>
			<width>200</width>
			<visible>!String.IsEmpty($PARAM[infolabel_prefix]ListItem.Premiered)</visible>
			<include content="InfoFlag">
				<param name="icon" value="lists/year.png" />
				<param name="label" value="$INFO[$PARAM[infolabel_prefix]ListItem.Premiered]" />
			</include>
		</control>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.Duration)" />
			<param name="flag_label" value="$INFO[ListItem.Duration]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.IsStereoscopic)" />
			<param name="flag_label" value="3D" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.VideoCodec)" />
			<param name="flag_label" value="$VAR[VideoListCodecVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.VideoResolution)" />
			<param name="flag_label" value="$INFO[ListItem.VideoResolution]$VAR[VideoListResolutionTypeVar, ]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.HdrType)" />
			<param name="flag_label" value="$VAR[VideoListHDRVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.VideoAspect)" />
			<param name="flag_label" value="$VAR[VideoListAspectVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.AudioCodec)" />
			<param name="flag_label" value="$VAR[AudioListCodecVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.AudioChannels)" />
			<param name="flag_label" value="$VAR[AudioListChannelsVar]" />
		</include>
	</include>
			</include>
		</control>
	</include>
	<include name="MediaFlagsLibrary">
		<visible>$PARAM[visible]</visible>
		<control type="group">
			<width>150</width>
			<visible>System.AddonIsEnabled(resource.images.studios.white) + !String.IsEmpty($PARAM[infolabel_prefix]ListItem.Studio)</visible>
			<include content="MediaFlag">
				<param name="texture" value="$INFO[$PARAM[infolabel_prefix]ListItem.Studio,resource://resource.images.studios.white/,.png]" />
			</include>
		</control>
		<control type="group">
			<top>10</top>
			<width>200</width>
			<visible>!String.IsEmpty($PARAM[infolabel_prefix]ListItem.Premiered)</visible>
			<include content="InfoFlag">
				<param name="icon" value="lists/year.png" />
				<param name="label" value="$INFO[$PARAM[infolabel_prefix]ListItem.Premiered]" />
			</include>
		</control>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.Duration)" />
			<param name="flag_label" value="$INFO[ListItem.Duration]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.IsStereoscopic)" />
			<param name="flag_label" value="3D" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.VideoCodec)" />
			<param name="flag_label" value="$VAR[VideoListCodecVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.VideoResolution)" />
			<param name="flag_label" value="$INFO[ListItem.VideoResolution]$VAR[VideoListResolutionTypeVar, ]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.HdrType)" />
			<param name="flag_label" value="$VAR[VideoListHDRVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.VideoAspect)" />
			<param name="flag_label" value="$VAR[VideoListAspectVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.AudioCodec)" />
			<param name="flag_label" value="$VAR[AudioListCodecVar]" />
		</include>
		<include content="MediaFlag">
			<param name="flag_visible" value="!String.IsEmpty(ListItem.AudioChannels)" />
			<param name="flag_label" value="$VAR[AudioListChannelsVar]" />
		</include>
	</include>
	<include name="MediaFlagsVideoPlayer">
		<visible>$PARAM[visible]</visible>
		<etc...
```

File: addons/skin.estuary/xml/Variables.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Variables.xml:812]: I went with what I thought would be more user friendly for those, instead of keeping it consistant in this case. I have a lot of old british shows on DVD and they all have 4:3 on the back, never personally seen 1.33 used anywhere. For 1.78 I do have DVD's with that on the back, but by far the majority use 16:9.

I don't have any hard conviction in doing these aspects this way, so if others prefer to keep it consistant I'm ok with that. At least it being in a xml now makes it editable.

File: addons/skin.estuary/xml/Includes.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Includes.xml]: Fixed

File: addons/skin.estuary/xml/Includes.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Includes.xml]: All traces of this should have been removed, this is now fixed

**Comment by jjd-uk** [comment]
> Not sure if better or worse but what about something like the below?

This would only give a very marginal line saving, so personally I don't believe the extra complexity to be worth it on this occasion.

File: addons/skin.estuary/xml/Variables.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Variables.xml]: fixed

**Comment by jjd-uk** [comment]
@MikeSiLVO all your comments addressed.

I'm not 100% sure if consistency is more desired than user friendliness. Normally I'd say consistency is best, however regarding the changes to the video aspects flags, to show 4:3 instead of 1.33:1, and 16:9 instead of 1.78:1 so I'd like to hear more views on this.

File: addons/skin.estuary/xml/Variables.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Variables.xml:812]: Oh and yes it could be simplified exactly as you show. I did have it exactly like that in a early version, however one of my aims was to make everything user editable, hence going with the full list to make it easier for any user wishing to do this.

Comment (jurialmunkey) [reply addons/skin.estuary/xml/Variables.xml:812]: I think consistency is important here. 

If you're going for screen size (16:9) over film aspect (1.78) as being more user friendly, then you should convert aspects to screen sizes for all common sizes not just 4:3 and 16:9.

Film aspect has the advantage of common denominator to compare easily -- e.g. I know from 2.33 vs 1.78 which is wider, but not necessarily between 2.33:1 and 16:9.

It's difficult because I think film aspect is more technically correct *for things shot on film*, whereas screen size would be more correct for things shot on video (e.g. for TV -- which is why you also see it on DVD and VHS where film has been converted to video and cropped for a smaller format).

Comment (jjd-uk) [reply addons/skin.estuary/xml/Variables.xml:812]: Ok that's me persuaded to keep consistency. I will revert back to an earlier simpler version of code that I had for the Aspect Ratio flags that didn't use the variables at all.

Comment (jjd-uk) [reply addons/skin.estuary/xml/Variables.xml]: Updated to the simpler method I have in a earlier version of this work.

**Review by Hitcher — APPROVED** [review]

File: addons/skin.estuary/xml/Variables.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Variables.xml]: Removed `HDR`

**Comment by jjd-uk** [comment]
I believe all comments have now be addressed. I will look to merge this sometime after 20th Sept if no objections received before then, so that gives another 7 days for any further comments.

File: addons/skin.estuary/xml/DialogSeekBar.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/DialogSeekBar.xml:73]: Just curious but why change seeklabel?

File: addons/skin.estuary/xml/Includes.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Includes.xml:399]: Just ListItem.IsStereoscopic?

File: addons/skin.estuary/xml/Includes.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Includes.xml:438]: Just VideoPlayer.IsStereoscopic?

File: addons/skin.estuary/xml/Variables.xml
Comment (MikeSiLVO) [inline addons/skin.estuary/xml/Variables.xml:955]: Entire VAR is for VideoPlayer or MusicPlayer?

**Comment by MikeSiLVO** [comment]
I think most of the VAR's can be consolidated but complexity and ease of modification was mentioned earlier so nothing against keeping it all the way it is :)

File: addons/skin.estuary/xml/DialogSeekBar.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/DialogSeekBar.xml:73]: To prevent overlap with the flags

File: addons/skin.estuary/xml/Includes.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Includes.xml:399]: Ah right, forgot it's a boolean.

File: addons/skin.estuary/xml/Includes.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/Includes.xml:438]: Same here, it's a boolean so need correcting.

File: addons/skin.estuary/xml/DialogSeekBar.xml
Comment (jjd-uk) [reply addons/skin.estuary/xml/DialogSeekBar.xml:73]: image

**Comment by jjd-uk** [comment]
Corrected incorrect boolean usage for the IsStereoscopic checks.

**Comment by jjd-uk** [comment]
> I think most of the VAR's can be consolidated but complexity and ease of modification was mentioned earlier so nothing against keeping it all the way it is :)

There's definitely scope for more consolidation. My original version had a single VAR for each flag type e.g. video codec containing both the ListItem and VideoPlayer codec checks with !Window.IsActive(fullscreeninfo) and Window.IsActive(fullscreeninfo) checks. But I decided to simplify and do seperate VAR's for the various ListItem & VidePlayer items at the expense of more VAR's.

What would have been really useful, but we don't have as skinners, is the ability to pass a Parameter into a Variable to allow something like:

`param name="prefix" value="ListItem" />` or `param name="prefix" value="VideoPlayer" />`

so in the VAR you could have something like:

`<value condition="String.IsEqual($PARM[prefix].VideoCodec,hevc)">H.265</value>`

Hopefully there is some C++ dev reading that could do something about that :)

**Comment by jjd-uk** [comment]
Final bump for comment before it gets merged merged either tomorrow or Sunday.

**Pull request merged**
